### PR TITLE
Add replyPrefixer client & message option

### PR DESCRIPTION
**Pull request by Gawdl3y**
**Please describe the changes this PR makes and why it should be merged:**
This adds the ability to override the default mention-based reply behaviour for Message#reply and supplying the reply option to TextChannel#send.

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by Drahcirius** [comment]
I think it might be useful if you added an additional parameter for the message object or even just pass the message object itself instead of the author so that you could use channel-specific or guild-specific context as a reply.

**Comment by Gawdl3y** [comment]
The message object isn't going to always be available. This function relies purely on the input from MessageOptions#reply, which can be an ID, message, member, or user.

**Comment by Drahcirius** [comment]
I think it might be useful to have a second parameter anyway, just so that you can add some checks and do something different depending on the context. 
e.g.
```js
replyPrefixer: (user, param) => {
  if (param.member) return `${param.member.displayName}: `;
  return `${user.username}: `;
}
```

**Comment by ItsPapaia** [comment]
> I think it might be useful to have a second parameter anyway, just so that you can add some checks and do something different depending on the context.

In that example you can just check if `user instanceof GuildMember` and process accordingly.

**Comment by HarmoGlace** [comment]
Will it be merged ?

**Comment by tguichaoua** [comment]
Could be usefull to resolve the user to have access to the client in the `ReplyPrefixer`.
```ts
  type ReplyPrefixer = (user: UserResolvable, client: Client) => string;
```

**Comment by NotSugden** [comment]
> Could be usefull to resolve the user to have access to the client in the `ReplyPrefixer`.
> 
> ```ts
>   type ReplyPrefixer = (user: UserResolvable, client: Client) => string;
> ```

seems a bit silly to me, why not resolve the user then pass it into the function,
```ts
type ReplyPrefixer = (user: User | GuildMember) => string
```

File: src/structures/APIMessage.js [hunk 1]
```diff
@@ -108,25 +108,27 @@ class APIMessage {
...
       const id = this.target.client.users.resolveID(this.options.reply);
-      mentionPart = `<@${this.options.reply instanceof GuildMember && this.options.reply.nickname ? '!' : ''}${id}>, `;
-      if (isSplit) {
-        splitOptions.prepend = `${mentionPart}${splitOptions.prepend || ''}`;
-      }
+      replyPrefix = this.options.replyPrefixer
+        ? this.options.replyPrefixer(this.options.reply)
+        : this.target.client.options.replyPrefixer
+        ? this.target.client.options.replyPrefixer(this.options.reply)
+        : `<@${this.options.reply instanceof GuildMember && this.options.reply.nickname ? '!' : ''}${id}>, `;
```
Comment (NotSugden) [inline src/structures/APIMessage.js:118]:
```suggestion
      const prefixer = this.options.replyPrefixer || this.target.client.options.replyPrefixer;
      replyPrefix = prefixer
        ? prefixer(this.options.reply)
        :  `<@${this.options.reply instanceof GuildMember && this.options.reply.nickname ? '!' : ''}${id}>, `;
```
a bit cleaner, easier to follow

**Review by MatievisTheKat — APPROVED** [review]

**Comment by almostSouji** [comment]
Considering discord uses the "reply" terminology for inline replies now and we're going ahead with removing the mention prefix functionality entirely to use proper inline replies in that name space instead (#4874) I'm going to close this.

**Pull request closed without merging**